monitoring/grafana/APIServerPriorityFairness.json

```diff
@@ -0,0 +1,845 @@
+        "target": {
+          "limit": 100,
+          "matchAny": false,
+          "tags": [],
+          "type": "dashboard"
+        },
+        "type": "dashboard"
+      }
+    ]
+  },
+  "description": "Dashboards for obsessive people",
+  "editable": true,
+  "fiscalYearStartMonth": 0,
+  "graphTooltip": 0,
+  "id": 4,
+  "links": [],
+  "liveNow": false,
+  "panels": [
+    {
+      "datasource": {
+        "default": true,
+        "type": "prometheus",
+        "uid": "prometheus-test"
+      },
+      "description": "Request rate by error code",
+      "fieldConfig": {
+        "defaults": {
+          "color": {
+            "mode": "palette-classic"
+          },
+          "custom": {
+            "axisBorderShow": false,
+            "axisCenteredZero": false,
+            "axisColorMode": "text",
+            "axisGridShow": false,
+            "axisLabel": "",
+            "axisPlacement": "auto",
+            "barAlignment": 0,
+            "barWidthFactor": 0.6,
+            "drawStyle": "line",
+            "fillOpacity": 58,
+            "gradientMode": "none",
+            "hideFrom": {
+              "legend": false,
+              "tooltip": false,
+              "viz": false
+            },
+            "insertNulls": false,
+            "lineInterpolation": "smooth",
+            "lineWidth": 8,
+            "pointSize": 5,
+            "scaleDistribution": {
+              "type": "linear"
+            },
+            "showPoints": "never",
+            "spanNulls": false,
+            "stacking": {
+              "group": "A",
+              "mode": "none"
+            },
+            "thresholdsStyle": {
+              "mode": "off"
+            }
+          },
+          "mappings": [],
+          "thresholds": {
+            "mode": "absolute",
+            "steps": [
+              {
+                "color": "light-blue",
+                "value": null
+              },
+              {
+                "color": "light-yellow",
+                "value": 80
+              }
+            ]
+          },
+          "unit": "none"
+        },
+        "overrides": []
+      },
+      "gridPos": {
+        "h": 12,
+        "w": 24,
+        "x": 0,
+        "y": 0
+      },
+      "id": 64,
+      "options": {
+        "legend": {
+          "calcs": [
+            "lastNotNull",
+            "max"
+          ],
+          "displayMode": "table",
+          "placement": "bottom",
+          "showLegend": true
+        },
+        "tooltip": {
+          "mode": "multi",
+          "sort": "desc"
+        }
+      },
+      "pluginVersion": "8.4.6",
+      "targets": [
+        {
+          "datasource": {
+            "type": "prometheus",
+            "uid": "$datasource"
+          },
+          "editorMode": "builder",
+          "exemplar": false,
+          "expr": "sum by (instance)(max_over_time(apiserver_current_inflight_requests{job=\"$APIService\", namespace=\"$SpaceNS\"}[$__rate_interval]))",
```

Comment:
`apiserver_current_inflight_requests` has an inherent sampling problem. See its description in https://kubernetes.io/docs/reference/instrumentation/metrics/ :

> Maximal number of currently used inflight request limit of this apiserver per request kind in last second

So unless the Prometheus server is scraping this metric every second, data is going missing.

Every metric that reports on utilization over a fixed recent period of time has this problem: unless the metric is scraped frequently enough (which is usually not true), data is lost.

There are other metrics that take a different approach. An illustrative example is the `apiserver_flowcontrol_read_vs_write_current_requests` histogram. It is a histogram of observations made once per nanosecond. The description follows.

> Observations, at the end of every nanosecond, of the number of requests (as a fraction of the relevant limit) waiting or in regular stage of execution.

Knowing the `--max-requests-inflight` (which actually is only about the readonly requests) and `--max-mutating-requests-inflight` enables you to convert those fractions to absolute numbers.

Since a histogram is a collection of counters, you can take the difference between any two scrapes taken at any two times and get all the data about the all the changes between those two times (albeit coarsened to the histogram bins, except for the `_sum` and `_count`, of course).